File: src/web/src/types/inventory.ts
import { InventorySourceSummaryDto } from './inventory-source';

// Legacy SourceType - kept for backwards compatibility
export type SourceType = 'Store' | 'Online' | 'Found' | 'Gift' | 'Trade' | 'Other';
export type InventoryCondition = 'Raw' | 'PreShaped' | 'Tumbled' | 'Polished' | 'Mixed';
export type InventoryStatus = 'Available' | 'InUse' | 'Depleted' | 'Partial';
export type SizeCategory = 'ZeroToOne' | 'OneToTwo' | 'TwoToThree' | 'ThreeToFour' | 'FourToFive' | 'GreaterThanFive' | 'Assorted';

export interface InventoryDto {
  inventoryId: string;
  name: string;
  acquiredDate: string;
  inventorySourceId: string | null;
  inventorySource: InventorySourceSummaryDto | null;
  totalWeightGrams: number | null;
  remainingWeightGrams: number | null;
  displayUnit: string;
  cost: number | null;  // Aggregated from specimens
  sizeCategories: SizeCategory[] | null;  // Aggregated from specimens
  qualityRating: number | null;  // Aggregated from specimens (average)
  storageLocation: string | null;
  notes: string | null;
  dateCreated: string;
  dateUpdated: string;
  specimens: InventorySpecimenDto[];
  photos: InventoryPhotoDto[];
  // Computed fields
  displayTotalWeight: number | null;
  displayRemainingWeight: number | null;
  photoCount: number;
  // Legacy fields for backwards compatibility
  sourceType: SourceType;
  sourceName: string | null;
  sourceLocation: string | null;
  sourceUrl: string | null;
  status: InventoryStatus;
  isFavorite: boolean;
}

export interface InventoryListDto {
  inventoryId: string;
  name: string;
  acquiredDate: string;
  inventorySourceId: string | null;
  sourceType: string | null;  // From InventorySource
  sourceName: string | null;  // From InventorySource.Name
  totalWeightGrams: number | null;
  remainingWeightGrams: number | null;
  displayUnit: string;
  cost: number | null;  // Aggregated from specimens
  qualityRating: number | null;  // Aggregated from specimens (average)
  dateCreated: string;
  specimenCount: number;
  photoCount: number;
  coverPhotoUrl: string | null;
  coverPhotoThumbnailUrl: string | null;
  // Specimen status counts
  availableCount: number;
  inUseCount: number;
  depletedCount: number;
  // Legacy fields for backwards compatibility
  status: InventoryStatus;
  isFavorite: boolean;
}

export interface CreateInventoryRequest {
  name: string;
  acquiredDate: string;
  inventorySourceId?: string;
  displayUnit?: string;
  storageLocation?: string;
  notes?: string;
  specimens?: CreateInventorySpecimenRequest[];
  // Legacy fields for backwards compatibility
  sourceType?: SourceType;
  sourceName?: string;
  sourceLocation?: string;
  sourceUrl?: string;
  status?: InventoryStatus;
  isFavorite?: boolean;
}

export interface UpdateInventoryRequest {
  name: string;
  acquiredDate: string;
  inventorySourceId?: string;
  displayUnit?: string;
  storageLocation?: string;
  notes?: string;
  specimens?: CreateInventorySpecimenRequest[];
  // Legacy fields for backwards compatibility
  sourceType?: SourceType;
  sourceName?: string;
  sourceLocation?: string;
  sourceUrl?: string;
  status?: InventoryStatus;
  isFavorite?: boolean;
}

export interface UpdateInventoryStatusRequest {
  status: InventoryStatus;
}

export interface UpdateInventorySpecimensRequest {
  specimens: CreateInventorySpecimenRequest[];
}

export interface InventorySpecimenDto {
  inventorySpecimenId: string;
  specimenId: string | null;
  userSpecimenId: string | null;
  commonName: string;
  scientificName: string | null;
  materialType: string;
  tumblingDifficulty: string | null;
  weightGrams: number | null;
  cost: number | null;
  condition: InventoryCondition | null;
  qualityRating: number | null;
  sizeCategories: SizeCategory[] | null;
  notes: string | null;
  status: InventoryStatus;
  storageLocation: string | null;
  url: string | null;
  source: 'system' | 'user';
}

export interface CreateInventorySpecimenRequest {
  specimenId?: string;
  userSpecimenId?: string;
  weightGrams?: number;
  cost?: number;
  condition?: InventoryCondition;
  qualityRating?: number;
  sizeCategories?: SizeCategory[];
  notes?: string;
  status?: InventoryStatus;
  storageLocation?: string;
  url?: string;
}

export interface InventoryPhotoDto {
  inventoryPhotoId: string;
  url: string;
  fileName: string | null;
  caption: string | null;
  isCover: boolean;
  sortOrder: number;
  dateCreated: string;
  thumbnailUrl: string | null;
  mediumUrl: string | null;
  largeUrl: string | null;
  blurHash: string | null;
  width: number | null;
  height: number | null;
  processingStatus: 'Processing' | 'Completed' | 'Failed';
  processingError: string | null;
  // Specimen link
  inventorySpecimenId: string | null;
  specimenName: string | null;
}

export interface InventoryStatsDto {
  totalItems: number;
  totalWeightGrams: number;
  totalInvested: number;
  availableCount: number;
  inUseCount: number;
  depletedCount: number;
}

export interface UploadInventoryPhotoResponse {
  success: boolean;
  photo?: InventoryPhotoDto;
  error?: string;
}

// Filter options for inventory list
export interface InventoryFilters {
  status?: InventoryStatus;
  sourceType?: SourceType;
  specimenId?: string;
  favorites?: boolean;
  search?: string;
  sortBy?: 'acquiredDate' | 'name' | 'weight' | 'cost' | 'dateCreated';
  sortOrder?: 'asc' | 'desc';
}

File: src/api/MyUglyRocks.Core/Entities/Inventory.cs
namespace MyUglyRocks.Core.Entities;

public enum SourceType
{
    Store = 0,      // Physical retail store
    Online = 1,     // Online purchase
    Found = 2,      // Collected in the wild
    Gift = 3,       // Received as gift
    Trade = 4,      // Traded with another collector
    Other = 5       // Other source
}

public enum InventoryCondition
{
    Raw = 0,        // Unprocessed, as found/purchased
    PreShaped = 1,  // Pre-shaped but not tumbled
    Tumbled = 2,    // Already tumbled (buying finished rocks)
    Polished = 3,   // Fully polished
    Mixed = 4       // Mix of conditions
}

public enum InventoryStatus
{
    Available = 0,  // Ready to use
    InUse = 1,      // Currently being used in active cycle(s)
    Depleted = 2,   // All used up
    Partial = 3     // Some used, some available
}

public enum SizeCategory
{
    ZeroToOne = 0,      // 0 - 1"
    OneToTwo = 1,       // 1" - 2"
    TwoToThree = 2,     // 2" - 3"
    ThreeToFour = 3,    // 3" - 4"
    FourToFive = 4,     // 4" - 5"
    GreaterThanFive = 5,// Greater than 5"
    Assorted = 6        // Assorted
}

public class Inventory : SoftDeletableEntity
{
    public Guid InventoryId { get; set; }
    public Guid UserId { get; set; }
    public Guid? InventorySourceId { get; set; }
    public required string Name { get; set; }
    public DateOnly AcquiredDate { get; set; }
    public decimal? TotalWeightGrams { get; set; }
    public decimal? RemainingWeightGrams { get; set; }
    public string DisplayUnit { get; set; } = "g";
    public decimal? Cost { get; set; }
    public string? SizeCategories { get; set; }  // Comma-separated list of SizeCategory values (aggregated from specimens)
    public int? QualityRating { get; set; } // 1-5
    public string? StorageLocation { get; set; }
    public string? Notes { get; set; }

    // Legacy fields - kept for migration, will be removed after migration
    [Obsolete("Use InventorySource.SourceType instead. Kept for migration.")]
    public SourceType SourceType { get; set; }
    [Obsolete("Use InventorySource.Name instead. Kept for migration.")]
    public string? SourceName { get; set; }
    [Obsolete("Use InventorySource.Location instead. Kept for migration.")]
    public string? SourceLocation { get; set; }
    [Obsolete("Use InventorySource.Url instead. Kept for migration.")]
    public string? SourceUrl { get; set; }
    [Obsolete("Status moved to InventorySpecimen. Kept for migration.")]
    public InventoryStatus Status { get; set; } = InventoryStatus.Available;
    [Obsolete("IsFavorite feature removed. Kept for migration.")]
    public bool IsFavorite { get; set; }

    // Navigation properties
    public virtual User User { get; set; } = null!;
    public virtual InventorySource? InventorySource { get; set; }
    public virtual ICollection<InventorySpecimen> InventorySpecimens { get; set; } = [];
    public virtual ICollection<InventoryPhoto> InventoryPhotos { get; set; } = [];
}
